### ISSUE TITLE
import strangeness with bare words

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5461]
Migrated from [rt.perl.org#9332](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9332) (status was 'resolved')

Searchable as RT9332$

**Comment by p5pRT** [comment]
### From liz@mickey.dijkmat.nl

#### Created by ELIZABETH@cpan.org



Three files are involved&#8203;:
=== Base\.pm =====================================================
package Base;

use strict;
@&#8203;Base&#8203;::ISA = qw\(\);
$Base&#8203;::VERSION = '0\.01';

1;

sub import \{

&nbsp;  my $class = shift;
warn "class = $class\, caller = "\.caller\(\);
warn "parameters = @&#8203;\_";
\} \#import
=================================================================

=== Actual\.pm ===================================================
package Actual;

use strict;
@&#8203;Actual&#8203;::ISA = qw\(Base\);
$Actual&#8203;::VERSION = '0\.01';

use Base \(bareword\); \# note bareword here

1;
=================================================================

=== test ========================================================
BEGIN \{
&nbsp;  use Carp \(\);
&nbsp;  $SIG\{\_\_DIE\_\_\} = \\&Carp&#8203;::confess;
\}

use Actual;
=================================================================

In perl 5\.6\.0 \(and possibly earlier\) it generates the following output&#8203;:

 class = Base\, caller = Actual at Base\.pm line 12\.
 parameters = bareword at Base\.pm line 13\.
 class = Actual\, caller = main at Base\.pm line 12\.
 parameters =  at Base\.pm line 13\.


Under Perl 5\.6\.1 and higher \(including bleadperl as far as I can see\)\,
executing "test" produces the following output&#8203;:

 class = Base\, caller = Actual at Base\.pm line 12\.
 parameters = bareword at Base\.pm line 13\.
 Unknown error
 Compilation failed in require at test line 6\.
&nbsp;     main&#8203;::BEGIN\(\) called at Actual\.pm line 6
&nbsp;     eval \{\.\.\.\} called at Actual\.pm line 6
 BEGIN failed\-\-compilation aborted at test line 6\.


The error disappears if you take out the \-use strict\- in Actual\.pm\.

So the problem appears to be related to the bareword in Actual\.pm causing
an error \(which it should\, because of the \-use strict\-\)\, but then not
being correctly handled afterwards\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.1:

Configured by root at Tue Jan 29 22:38:37 CET 2002.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.7-10, archname=i686-linux
    uname='linux ds051.xs4all.nl 2.4.7-10 #1 thu sep 6 17:27:27 edt 2001 i686 unknown '
    config_args='-de'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-98)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/i686-linux
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/liz
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

liz@&#8203;mickey\.dijkmat\.nl wrote&#8203;:
\[\.\.\.\]

The bug can be cut down to this :

Actual\.pm :
&nbsp;   package Actual;
&nbsp;   use strict;
&nbsp;   use less bareword;
&nbsp;   1;

$ bleadperl \-MActual \-e1
Unknown error
Compilation failed in require\.
BEGIN failed\-\-compilation aborted\.

> The error disappears if you take out the \-use strict\- in Actual\.pm\.
> 
> So the problem appears to be related to the bareword in Actual\.pm causing
> an error \(which it should\, because of the \-use strict\-\)\, but then not
> being correctly handled afterwards\.

Yes\, the fact that perl issues an error is OK\, but the error
message is misleading; it must be lost somewhere\.

\-\- 
Rafael Garcia\-Suarez

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Tue\, 14 May 2002 12&#8203;:27&#8203;:32 \+0200\, Rafael Garcia\-Suarez \<raphel\.garcia\-suarez@&#8203;hexaflux\.com> said&#8203;:

&nbsp;  > liz@&#8203;mickey\.dijkmat\.nl wrote&#8203;:
&nbsp;  > \[\.\.\.\]

&nbsp;  > The bug can be cut down to this :

&nbsp;  > Actual\.pm :
&nbsp;  >    package Actual;
&nbsp;  >    use strict;
&nbsp;  >    use less bareword;
&nbsp;  >    1;

&nbsp;  > $ bleadperl \-MActual \-e1
&nbsp;  > Unknown error
&nbsp;  > Compilation failed in require\.
&nbsp;  > BEGIN failed\-\-compilation aborted\.

I was just about to write a bug report that is most probably connected
to the same bug but maybe more\. We have an exception being thrown in
an eval but no $@&#8203; being set\.


% /usr/local/perl\-5\.7\.3@&#8203;16589/bin/perl  \-e '
use strict; our $testno;
our $eval1 = q\{  BEGIN \{ use strict bareword; $&#8203;::X1=1; \}  \};
eval $eval1;
printf "not " unless $@&#8203;;
print "ok "\, \+\+$testno \," \# eval1\\n";
printf "not " if $&#8203;::X1;
print "ok "\, \+\+$testno \," \# X1\\n";
our $eval2 = q\{  use strict bareword; $&#8203;::X2=1; \};
eval $eval2;
printf "not " unless $@&#8203;;
print "ok "\, \+\+$testno \," \# eval2\\n";
printf "not " if $&#8203;::X2;
print "ok "\, \+\+$testno \," \# X2\\n";
'
ok 1 \# eval1
ok 2 \# X1
not ok 3 \# eval2
ok 4 \# X2



\-\- 
andreas

**Comment by p5pRT** [comment]
### From @lizmat

At 01&#8203;:21 PM 5/14/02 \+0200\, Andreas J\. Koenig wrote&#8203;:
> >>>>> On Tue\, 14 May 2002 12&#8203;:27&#8203;:32 \+0200\, Rafael Garcia\-Suarez 
> \<raphel\.garcia\- > The bug can be cut down to this :
>   > Actual\.pm :
>   >    package Actual;
>   >    use strict;
>   >    use less bareword;
>   >    1;

Just for the record and emphasis\.  5\.6\.1 is already suffering from this 
problem\, 5\.6\.0 does not give an error message&#8203;: it just happily plods along\.\.\.


Elizabeth Mattijsen

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Tue\, 14 May 2002 13&#8203;:43&#8203;:06 \+0200\, Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl> said&#8203;:

&nbsp;  > At 01&#8203;:21 PM 5/14/02 \+0200\, Andreas J\. Koenig wrote&#8203;:
 >> >>>>> On Tue\, 14 May 2002 12&#8203;:27&#8203;:32 \+0200\, Rafael Garcia\-Suarez
 >> \<raphel\.garcia\- > The bug can be cut down to this :
 >> > Actual\.pm :
 >> >    package Actual;
 >> >    use strict;
 >> >    use less bareword;
 >> >    1;

&nbsp;  > Just for the record and emphasis\.  5\.6\.1 is already suffering from
&nbsp;  > this problem\, 5\.6\.0 does not give an error message&#8203;: it just happily
&nbsp;  > plods along\.\.\.

Thanks for the reminder\! Archaeology proves that my bug and your bug
are not the same\. "Your" bug started with 5\.005\_62 to silently ignore
the bareword\, and with patch 5974 it started to emit the "Unknown
error" message\. "My" bug seems is much older\.


\-\- 
andreas

**Comment by p5pRT** [comment]
### From @lizmat

At 03&#8203;:50 PM 5/14/02 \+0200\, Andreas J\. Koenig wrote&#8203;:
> >>>>> On Tue\, 14 May 2002 13&#8203;:43&#8203;:06 \+0200\, Elizabeth Mattijsen > Just for 
> the record and emphasis\.  5\.6\.1 is already suffering from
>   > this problem\, 5\.6\.0 does not give an error message&#8203;: it just happily
>   > plods along\.\.\.
>Thanks for the reminder\! Archaeology proves that my bug and your bug
>are not the same\. "Your" bug started with 5\.005\_62 to silently ignore
>the bareword\, and with patch 5974 it started to emit the "Unknown
>error" message\.

I don't think the bareword is ignored\.  As far as I can see\, the error 
occurs upon \_returning\_ from the import\(\) routine\, which \_is\_ called 
correctly \(at least in 5\.6\.0\, 5\.6\.1 and bleadperl as of 30 April\)\.\.\.


Elizabeth Mattijsen

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Tue\, 14 May 2002 16&#8203;:22&#8203;:44 \+0200\, Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl> said&#8203;:

 >> Thanks for the reminder\! Archaeology proves that my bug and your bug
 >> are not the same\. "Your" bug started with 5\.005\_62 to silently ignore
 >> the bareword\, and with patch 5974 it started to emit the "Unknown
 >> error" message\.

&nbsp;  > I don't think the bareword is ignored\.  As far as I can see\, the error
&nbsp;  > occurs upon \_returning\_ from the import\(\) routine\, which \_is\_ called
&nbsp;  > correctly \(at least in 5\.6\.0\, 5\.6\.1 and bleadperl as of 30 April\)\.\.\.

Sorry\, I didn't express myself correctly\. I meant 5\.005\_62 started to
silently ignore the violation of the strict pragma\.

To sum up my findings without \(I hope&#8203;:\-\) ambiguity&#8203;:

% for p in /usr/local/perl5\.005\_61 /usr/local/perl5\.005\_62 /usr/local/perl\-5\.6\.0@&#8203;5973 /usr/local/perl\-5\.6\.0@&#8203;5974   ; do 
echo $p
$p/bin/perl tests/test\_useactual\.pl
done
/usr/local/perl5\.005\_61
Bareword "bareword" not allowed while "strict subs" in use at Actual\.pm line 3\.
/usr/local/perl5\.005\_62
/usr/local/perl\-5\.6\.0@&#8203;5973
/usr/local/perl\-5\.6\.0@&#8203;5974
Unknown error
Compilation failed in require at tests/test\_useactual\.pl line 1\.
BEGIN failed\-\-compilation aborted at tests/test\_useactual\.pl line 1\.



\-\- 
andreas

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

Original code from the report&#8203;:

perl\-5\.6\.1 rt\-9332\.pl
class = Base\, caller = Actual at /tmp/rt\-9332//Base\.pm line 12\.
parameters = bareword at /tmp/rt\-9332//Base\.pm line 13\.
Unknown error
Compilation failed in require at rt\-9332\.pl line 10\.
&nbsp;        main&#8203;::BEGIN\(\) called at /tmp/rt\-9332//Actual\.pm line 10
&nbsp;        eval \{\.\.\.\} called at /tmp/rt\-9332//Actual\.pm line 10
BEGIN failed\-\-compilation aborted at rt\-9332\.pl line 10\.

perl\-5\.8\.8 rt\-9332\.pl
class = Base\, caller = Actual at /tmp/rt\-9332//Base\.pm line 12\.
parameters = bareword at /tmp/rt\-9332//Base\.pm line 13\.
Unknown error
Compilation failed in require at rt\-9332\.pl line 10\.
 at rt\-9332\.pl line 10
&nbsp;        main&#8203;::BEGIN\(\) called at /tmp/rt\-9332//Actual\.pm line 10
&nbsp;        eval \{\.\.\.\} called at /tmp/rt\-9332//Actual\.pm line 10
BEGIN failed\-\-compilation aborted at rt\-9332\.pl line 10\.
 at rt\-9332\.pl line 10


perl\-5\.8\.9\-tobe rt\-9332\.pl
Bareword "bareword" not allowed while "strict subs" in use at /tmp/rt\-
9332//Actual\.pm line 7\.
Compilation failed in require at rt\-9332\.pl line 10\.
 at rt\-9332\.pl line 10
&nbsp;        main&#8203;::BEGIN\(\) called at /tmp/rt\-9332//Actual\.pm line 10
&nbsp;        eval \{\.\.\.\} called at /tmp/rt\-9332//Actual\.pm line 10
BEGIN failed\-\-compilation aborted at rt\-9332\.pl line 10\.
 at rt\-9332\.pl line 10


perl\-5\.10\.0 rt\-9332\.pl
Bareword "bareword" not allowed while "strict subs" in use at /tmp/rt\-
9332//Actual\.pm line 7\.
Compilation failed in require at rt\-9332\.pl line 10\.
 at rt\-9332\.pl line 10
&nbsp;        main&#8203;::BEGIN\(\) called at /tmp/rt\-9332//Actual\.pm line 10
&nbsp;        eval \{\.\.\.\} called at /tmp/rt\-9332//Actual\.pm line 10
BEGIN failed\-\-compilation aborted at rt\-9332\.pl line 10\.
 at rt\-9332\.pl line 10


Rafael's test case&#8203;:

perl\-5\.6\.1 \-MActual \-e1 
Unknown error
Compilation failed in require\.
BEGIN failed\-\-compilation aborted\.

perl\-5\.8\.8 \-MActual \-e1 
Unknown error
Compilation failed in require\.
BEGIN failed\-\-compilation aborted\.

perl\-5\.8\.9\-tobe \-MActual \-e1 
Bareword "bareword" not allowed while "strict subs" in use at /tmp/
Actual\.pm line 3\.
Compilation failed in require\.
BEGIN failed\-\-compilation aborted\.

perl\-5\.10\.0 \-MActual \-e1 
Bareword "bareword" not allowed while "strict subs" in use at /tmp/
Actual\.pm line 3\.
Compilation failed in require\.
BEGIN failed\-\-compilation aborted\.


Andreas's test case&#8203;:

perl\-5\.6\.1 rt\-9332\-2\.pl 
ok 1 \# eval1
ok 2 \# X1
not ok 3 \# eval2
ok 4 \# X2


perl\-5\.8\.0 rt\-9332\-2\.pl 
ok 1 \# eval1
ok 2 \# X1
not ok 3 \# eval2
ok 4 \# X2

perl\-5\.8\.1 rt\-9332\-2\.pl 
ok 1 \# eval1
ok 2 \# X1
ok 3 \# eval2
ok 4 \# X2

perl\-5\.10\.0 rt\-9332\-2\.pl 
ok 1 \# eval1
ok 2 \# X1
ok 3 \# eval2
ok 4 \# X2

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'
